### ISSUE TITLE
Change in dpi settings handling between 12.1 and 12.2?

### ISSUE THREAD

**Issue by dbo-odoo** [wkhtmltopdf/wkhtmltopdf#2156]
Hi!

Did you change the way dpi are handled recently? On Ubuntu, even though you specify that it shouldn't work on X11 systems, the dpi arg was working up to 12.1. It's not the case anymore (example in image - same magnification).
![selection_005](https://cloud.githubusercontent.com/assets/9530767/5817328/dddc20f8-a0aa-11e4-89f8-ec4fd7c26e10.png)

I was wondering if I could use zoom instead of dpi to achieve something comparable?

**Comment by ashkulz** [comment]
What is the `--dpi` value you were trying to use? The only relevant change is #1863 which kicks in when you specify a low DPI.

**Comment by dbo-odoo** [comment]
No, the dpi was not especially low: 90

**Comment by ashkulz** [comment]
That is the less than the default DPI on Windows (which is 96). Earlier it was not correct as it would ignore/crash it but now it respects the lower DPI (and effectively zooming in the page). I'd recommend you to use 96 instead (or choose an appropriate zoom value (`90/96`).

**Comment by sle-odoo** [comment]
@ashkulz 
Hello,
It's reproducible on GNU/Linux. I'm currently on Debian Jessie x64 and a command like `wkhtmltopdf --dpi 96 www.google.be a.pdf` behaves differently between 0.12.0 and 0.12.2.1: as expected in 0.12.0 and ignored on 0.12.2.1.
I tried with some specific values like 72, 90, 96, 1000 but it's looking like this argument has no effect at all.
Thanks in advance

**Comment by sle-odoo** [comment]
I'll add that i see the same behavior in Windows: `--dpi` looks like being ignored.
However, `--zoom` is working as expected on both platforms.
You should reopen this issue if you consider that it's a duplicate of the report of the crash on Windows when the DPI is lower than 96: it's not.

**Comment by ashkulz** [comment]
So does it look like a regression since `0.12.1`? Can you see if reverting wkhtmltopdf/qt@ccffd0d9842e06ed1521fa669faec2c50e6bae71 works better for you?

**Comment by sle-odoo** [comment]
@ashkulz reverting the [commit you pointed out](https://github.com/wkhtmltopdf/qt/commit/ccffd0d9842e06ed1521fa669faec2c50e6bae71) brings brack the `--dpi` behavior! `wkhtmltopdf --dpi 130 www.google.be a.pdf` works definitely as expected. So yes it is a regression since 0.12.1.

I tried it on my x64 Jessie and i must note that using low values (e.g. 30) result in a blank page and big value (e.g. 250) result in (i suppose) a fallback on the default dpi value. But it may already be the case with the last release. I'll confirm you this tomorrow, and i'll try to build and test it on Windows too.

To confirm this issue, run
`wkhtmltopdf www.google.be a.pdf`
`wkhtmltopdf --dpi 90 www.google.be a.pdf`
`wkhtmltopdf --dpi 130 www.google.be a.pdf`
with a build with and without the revert of the commit.

Can i do something more to help you fix this?

By the way, well done on the build script. I haven't tried it since before 0.12 and it really has improved a lot.

**Comment by ashkulz** [comment]
Well, the fix was proposed in wkhtmltopdf/qt#12 and eventually merged because that is the way the code is upstream. Can you post screenshots on what happens with various DPI levels with both `0.12.1` and `0.12.2.1`?

**Comment by sle-odoo** [comment]
Hi,
Here are screenshots on what happens with various DPI levels with both 0.12.1 and 0.12.2.1 (on jessie x64). As it's quite unreadable, here is a [link to an archive](http://nightly.odoo.com/tmp/wkhtmltopdf_dpi.tar.gz) with all the generated pdf.

I confirm that reverting the commit you pointed out brings back exactly the same behavior as in the 0.12.1 release.

**0.12.1**
**without dpi arg**
![0_12_1 pdf](https://cloud.githubusercontent.com/assets/7601090/5853481/6e72d5b0-a224-11e4-9b2c-5bce6ccd164e.png)
**with 72 as dpi arg**
![0_12_1_dpi72 pdf](https://cloud.githubusercontent.com/assets/7601090/5853486/7dedf358-a224-11e4-8450-554b58415a99.png)
**with 96 as dpi arg**
![0_12_1_dpi96 pdf](https://cloud.githubusercontent.com/assets/7601090/5853487/7deebde2-a224-11e4-9924-a96303b7120d.png)
**with 130 as dpi arg**
![0_12_1_dpi130 pdf](https://cloud.githubusercontent.com/assets/7601090/5853488/7df1eca6-a224-11e4-9b61-2ef1216409f0.png)

**0.12.2.1**
**without dpi arg**
![0_12_2_1 pdf](https://cloud.githubusercontent.com/assets/7601090/5853588/3bfb1632-a225-11e4-9732-b11b7cbf4ad3.png)
**with 72 as dpi arg**
![0_12_2_1_dpi72 pdf](https://cloud.githubusercontent.com/assets/7601090/5853589/3bfba3e0-a225-11e4-922d-27d094d1abab.png)
**with 96 as dpi arg**
![0_12_2_1_dpi96 pdf](https://cloud.githubusercontent.com/assets/7601090/5853590/3c0290d8-a225-11e4-859c-31a7a7c5bb39.png)
**with 130 as dpi arg**
![0_12_2_1_dpi130 pdf](https://cloud.githubusercontent.com/assets/7601090/5853591/3c0691ba-a225-11e4-851f-a8b5aa0ac7b8.png)

**Comment by RogerJoensson** [comment]
Hi,

I just downloaded and tried wkhtmltopdf. I get a resolution of 93.6 with the below html:
The 4 "D"s printed are supposed to be spaced 100mm at 96dpi, which I figured was what is was making (and not 600dpi as the manual says), but it seems to not do that either. I get about 93.6dpi and it also seems to vary slightly with what I input.

I am on mac os x 10.8.5 and I input this into the terminal:
Rogersmacmini:~ roger$ wkhtmltopdf --grayscale -L 0 -R 0 -T 0 -B 0 /Users/roger/Desktop/Ugga.html x.pdf
The dpi options have no effect (and should not?). Changing the margins sometimes changes the scaling too.

<!DOCTYPE html><html><body><p style="font-family:'Arial'; font-size:30px; color:#BDBDBD; position: absolute; top:1.795276px;left:37.79528px;">D</p></body></html>
<!DOCTYPE html><html><body><p style="font-family:'Arial'; font-size:30px; color:#BDBDBD; position: absolute; top:1.795276px;left:415.748px;">D</p></body></html>
<!DOCTYPE html><html><body><p style="font-family:'Arial'; font-size:30px; color:#BDBDBD; position: absolute; top:379.748px;left:37.79528px;">D</p></body></html>
<!DOCTYPE html><html><body><p style="font-family:'Arial'; font-size:30px; color:#BDBDBD; position: absolute; top:379.748px;left:415.748px;">D</p></body></html>

**Comment by vevisystems** [comment]
In Linux, with 12.1, setting dpi to 96 (even though it is reported not to work) and disabling smart shrinking is the only way I have found to achieve printed results that match what you specify on CSS. I mean if I specify width: 1cm in CSS it actually is 1cm width in the printed page. However, with 12.2.1 setting dpi to 96 seems to not work (it is the same as not setting dpi at all), and the results are "larger", this is, if you specify 1cm in PDF, printed size in larger.

¿Is there any way to achieve same results in 12.2.1 on a Linux machine?

**Comment by RogerJoensson** [comment]
Hm.

I using px.

If I do:

<table style="width:100%;font-family:'Arial';font-size:5mm; position: absolute; top:10mm; left:10mm; width:100mm; height:100mm"; cellpadding="0">
<style>table, th, td {border: 0.3mm solid black}</style>
</table>

I get a square that is 127mm.

**Comment by vevisystems** [comment]
We've been testing a while and it seems that removing the dpi parameter and setting zoom to 0.78125 (75/96) gets a correct print size, so this may be a workaround for anybody needing this.

**Comment by binarysource** [comment]
Having exact same issue as original post, @vevisystems (thanks!) suggestion of setting zoom to 0.78124 appears to have fix it temporarily.

**Comment by ashkulz** [comment]
should be fixed as #2463 is now merged.

**Comment by ashkulz** [comment]
[`0.12.4`](http://wkhtmltopdf.org/downloads.html#stable) has been released, which should contain the fix for this issue. Please report back if it is not solved with the above version.

**Comment by hakimio** [comment]
DPI setting still doesn't seem to have any effect when using latest stable build even on windows, but using zoom instead seems to do the trick.

**Comment by victor-ponamariov** [comment]
+1, Dpi does not work for me

**Comment by mgw-wjp** [comment]
+1, Dpi does not work for me

**Comment by JKE-be** [comment]
Hello,

Just tested with `wkhtmltopdf 0.12.4 (with patched qt)` and --dpi setting still doesn't seem to have any effect.

cc @sle-odoo

**Comment by danielson317** [comment]
+1 DPI Still does not work.

**Comment by danger89** [comment]
😢

**Comment by danger89** [comment]
I'm using version `0.12.4 (with patched qt)` as well, I do see some differences in the quality of the letters, although slightly. But not more. 

Are we all on the same side what the DPI setting does exactly? Some people expect the PDF result to be zoomed in as well with a different DPI setting, is that the correct assumption?

**Comment by danger89** [comment]
Will wkhtmltopdf version 0.12.5 fix the issue you think? https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3241 

Hopefully 0.12.5 will be released somewhere soon... Waiting on closing ticket https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2609 and hopefully 0.12.5 will be released.
